src/psyclone/domain/gocean/transformations/gocean_opencl_trans.py

```diff
@@ -243,29 +253,552 @@ def apply(self, node, options=None):
+            for child in callblock.pop_all_children():
+                node.children.insert(outerloop.position, child)
+
+            # Then we call the clEnqueueNDRangeKernel
+            assig = Assignment.create(
+                        Reference(flag),
+                        Call.create(cl_launch, [
+                            # OpenCL Command Queue
+                            cmd_queue,
+                            # OpenCL Kernel object
+                            Reference(kernelsym),
+                            # Number of work dimensions
+                            Literal("2", INTEGER_TYPE),
+                            # Global offset (if NULL the global IDs start at
+                            # offset (0,0,0))
+                            Reference(c_null),
+                            # Global work size
+                            Call.create(c_loc, [Reference(global_size)]),
+                            # Local work size
+                            Call.create(c_loc, [Reference(local_size)]),
+                            # Number of events in wait list
+                            Literal("0", INTEGER_TYPE),
+                            # Event wait list that need to be completed before
+                            # this kernel
+                            Reference(c_null),
+                            # Event that identifies this kernel completion
+                            Reference(c_null)]))
+            assig.preceding_comment = "Launch the kernel"
+            node.children.insert(outerloop.position, assig)
+            self._insert_kernel_code_in_opencl_file(kern)
+
+            # Add additional checks if we are in debug mode
+            if api_config.debug_mode:
+
+                # First check the launch return value
+                message = Literal(
+                    f"{kern.name} clEnqueueNDRangeKernel", CHARACTER_TYPE)
+                check = Call.create(check_status, [message, Reference(flag)])
+                node.children.insert(outerloop.position, check)
+
+                # Then add a barrier and check that the kernel executed
+                # successfully
+                barrier = Assignment.create(
+                            Reference(flag),
+                            Call.create(cl_finish, [cmd_queue.copy()]))
+                node.children.insert(outerloop.position, barrier)
+                message = Literal(f"Errors during {kern.name}", CHARACTER_TYPE)
+                check = Call.create(check_status, [message, Reference(flag)])
+                node.children.insert(outerloop.position, check)
+
+            nodes_to_detach.append(outerloop)
+
+        # If we execute the kernels asynchronously, we need to add wait
+        # statements before the PSy-layer operations that depend on them
+        for possible_dependent_node in node.walk(HaloExchange):
+            dependency = possible_dependent_node.backward_dependence()
```

Comment:
Probably worth a comment here saying that `backward_dependence` returns the last Loop that contained a kernel that updated blah blah.

Reply:
Done, I think